routingpy/base.py

```diff
@@ -0,0 +1,206 @@
+
+
+class BaseClient(metaclass=ABCMeta):
+    """Abstract base class every client inherits from. Authentication is handled in each subclass."""
+
+    def __init__(
+            self,
+            base_url,
+            user_agent=None,
+            timeout=DEFAULT,
+            retry_timeout=None,
+            retry_over_query_limit=None,
+            skip_api_error=None
+    ):
+        """
+        :param base_url: The base URL for the request. All routers must provide a default.
+            Should not have a trailing slash.
+        :type base_url: string
+
+        :param user_agent: User-Agent to send with the requests to routing API.
+            Overrides ``options.default_user_agent``.
+        :type user_agent: string
+
+        :param timeout: Combined connect and read timeout for HTTP requests, in
+            seconds. Specify "None" for no timeout.
+        :type timeout: int
+
+        :param retry_timeout: Timeout across multiple retriable requests, in
+            seconds.
+        :type retry_timeout: int
+
+        :param requests_kwargs: Extra keyword arguments for the requests
+            library, which among other things allow for proxy auth to be
+            implemented.
+
+            Example:
+
+            >>> from routingpy.routers import ORS
+            >>> router = ORS(my_key, requests_kwargs={'proxies': {'https': '129.125.12.0'}})
+            >>> print(router.proxies)
+            {'https': '129.125.12.0'}
+
+        :param retry_over_query_limit: If True, client will not raise an exception
+            on HTTP 429, but instead jitter a sleeping timer to pause between
+            requests until HTTP 200 or retry_timeout is reached.
+        :type retry_over_query_limit: bool
+
+        :param skip_api_error: Continue with batch processing if a :class:`routingpy.exceptions.RouterApiError` is
+            encountered (e.g. no route found). If False, processing will discontinue and raise an error. Default False.
+        :type skip_api_error: bool
+        """
+        self.base_url = base_url
+
+        self.retry_over_query_limit = retry_over_query_limit if retry_over_query_limit is False else options.default_retry_over_query_limit
+        self.retry_timeout = timedelta(seconds=retry_timeout or options.default_retry_timeout)
+
+        self.skip_api_error = skip_api_error or options.default_skip_api_error
+
+        self.headers = {
+            "User-Agent": user_agent or options.default_user_agent,
+            'Content-Type': 'application/json'
+        }
+
+        self.timeout = timeout if timeout != DEFAULT else options.default_timeout
+
+        self._req = None
+
+    @abstractmethod
+    def _request(self):
+        """Must be implemented for inheriting client classes."""
+        pass
+
+    @property
+    def req(self):
+        """Holds the :class:`requests.PreparedRequest` property for the last request."""
+        return self._req
+
+    @staticmethod
+    def _get_body(response):
```

Comment:
put this whole method into the default client, since it has requests specific logic on line 168